### ISSUE TITLE
Discussion: C# Break nested loop

### ISSUE THREAD

**Issue by jcouv** [dotnet/csharplang#869]
@Danthekilla commented on [Thu Aug 31 2017](https://github.com/dotnet/roslyn/issues/21864)

In C# it would be fanstastic to have a language feature (probably syntactic sugar) that lets you break nested loops.

Currently these are the 2 ways you can break out of nested loops:

```
var foundValue;
for (int x = 0; x < xMax; x++)
{
    for (int y = 0; y < yMax; y++)
    {
        foundValue = GetValue(x, y);
        if (foundValue == valueToCompare)
            goto ENDOFLOOPS;
    }
}
ENDOFLOOPS:
```

```
var foundValue;
bool shouldBreak;
for (int x = 0; x < xMax; x++)
{
    for (int y = 0; y < yMax; y++)
    {
        foundValue = GetValue(x, y);
        if (foundValue == valueToCompare)
        {
            shouldBreak = true;
            break;
        }
    }
    if (shouldBreak)
        break;
}
```

I would like to propose the ability for a break to have a value for how many loops to break out of. Like so:

```
var foundValue;
for (int x = 0; x < xMax; x++)
{
    for (int y = 0; y < yMax; y++)
    {
        foundValue = GetValue(x, y);
        if (foundValue == valueToCompare)
            break 2;
    }
}
```

Not only is it slightly smaller, but I think easier to read.

Anyone have any thoughts on the matter?

**Comment by DavidArno** [comment]
@Danthekilla,

You forgot the existing third way:
```cs
(bool found, ValueType value) Foo()
{
    for (int x = 0; x < xMax; x++)
    {
        for (int y = 0; y < yMax; y++)
        {
            var foundValue = GetValue(x, y);
            if (foundValue == valueToCompare) return (true, foundValue);
        }
    }
    return (false, default);
}
```

Small focused functions are your friend. But if you want long functions, as you say, there is already two ways of achieving this goal. So I'd see yet another way of achieving it as really bringing nothing to the language.

**Comment by jnm2** [comment]
Sometimes it's a loop and a switch and coming up with a name for it is super arbitrary and counter-intuitive.

**Comment by eyalsk** [comment]
I'm not in favour of this but I think the following is better `goto ENDOFLOOPS when foundValue == valueToCompare;` I know, no one "likes" the `goto` statement but anyway, we might get a `goto` and `continue` expressions so the following would be possible:

```
var foundValue;
for (int x = 0; x < xMax; x++)
{
    for (int y = 0; y < yMax; y++)
    {
        foundValue = GetValue(x, y);
	foundValue == valueToCompare ? goto ENDOFLOOPS : continue;
    }
}
ENDOFLOOPS:
```

**Comment by Logerfo** [comment]
I'd prefer to see something like [Java Branching Statements](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/branch.html), as once pointed out by @HaloFour  at https://github.com/dotnet/csharplang/issues/176#issuecomment-310874877

**Comment by jnm2** [comment]
`goto` is perfectly reasonable if you only go to points that can already be jumped to by another kind of branching statement like `break`. Same with `goto case`.

**Comment by eyalsk** [comment]
@Logerfo, Indeed, it makes things clear.

**Comment by benaadams** [comment]
Inline returns slow down loops until .NET Core 2.1 https://github.com/dotnet/coreclr/pull/13314 and functions with a loop generally won't inline (without forcing); so wouldn't be so great for very hot loops; though ok for general loops.

**Comment by bondsbw** [comment]
`break <number>` doesn't refactor as well.  Adding another loop or removing a loop requires carefully reading back through the code and reworking such statements.  Otherwise, it can result in jumping to an unintended point.

**Comment by jnm2** [comment]
@bondsbw Not to mention copy/paste.

**Comment by MuletTheGreat** [comment]
This is simple, short and elegant. Goto's require maintenance and their excessive use results in spaghetti code. Extra temp variables hanging around is just junk code that can go wrong.

Break = n; in an n deep loop would create short, concise code that cannot be confused for anything else.

**Comment by HaloFour** [comment]
I agree with some of the other posters, I don't like the use of an integer indicating the number of levels.  If any enhancements to `break` or `continue` are considered for nested loop scenarios I'd prefer the labeled loop syntax used by Java.

```java
String foundValue = null;
OUTERLOOP:
for (int x = 0; x < xMax; x++)
{
    for (int y = 0; y < yMax; y++)
    {
        foundValue = GetValue(x, y);
        if (foundValue == valueToCompare)
            break OUTERLOOP;
    }
}
```

This makes it much clearer exactly where the `break`/`continue` statement will continue execution.

**Comment by orthoxerox** [comment]
I agree that breaking from a named loop is the best option. I think there was a proposal of mine asking for that. Or was it left in the roslyn repo?

**Comment by eyalsk** [comment]
@jnm2 It's better to have an "arbitrary" label and know what it is than having a number and starting to wonder what it means, sometimes.

**Comment by jnm2** [comment]
@eyalsk Exactly my point. 😄

**Comment by eyalsk** [comment]
@jnm2 ah weird, I thought you wrote something about it. 😉

**Comment by jnm2** [comment]
@eyalsk I don't remember preferring numbers but I'm happy to update my opinion if I did write that somewhere. I'm not seeing it.

**Comment by eyalsk** [comment]
@jnm2 Well, I didn't say anything about preferring it, I'm pretty sure you wrote something and deleted it but if you're saying you didn't then I might have gotten the wrong person.

**Comment by eyalsk** [comment]
@jnm2 I was referring to this comment.

```
Sometimes it's a loop and a switch and coming up with a name for it is super arbitrary and counter-intuitive.
```

**Comment by jnm2** [comment]
Ah, I was answering @DavidArno's suggestion that you refactor it into a separate method which would then need a name. It did not escape me that you also need a name for a label, but it the label name is free to be coupled to only make sense in the context of the control flow in the method's implementation whereas refactoring it to a separate method requires it to stand on its own.

**Comment by eyalsk** [comment]
@jnm2 got'cha. 😉

**Comment by svick** [comment]
@jnm2 Doesn't that change when you use a local function instead of a method? I think that the name of a label and the name of a local function can both be specific to a single method, or even to a scope within a method.

**Comment by jnm2** [comment]
In theory. That's a good point.

**Comment by Danthekilla** [comment]
Personally I find that coming up with a name for a goto can be very arbitrary and quite counter-intuitive. 

Which is why I would prefer to use a 'break integer' syntax.

**Comment by jnm2** [comment]
Already today a goto label name like `exit_foolist_loop` is not arbitrary or counter-intuitive.

**Comment by eyalsk** [comment]
@Danthekilla 

> Personally I find that coming up with a name for a goto can be very arbitrary and quite counter-intuitive.

Yes, naming is hard, we all know that but just like we come up with names that make sense for other things throughout the code we should do the same here.

goto:
```
foundValue == valueToCompare ? goto end_of_find_item_loop : continue;
```

break:
```
foundValue == valueToCompare ? break find_item_loop : continue;
```

Readable code should be easy to follow and understand so doing something like `break <number>` might be easier to understand but not follow as such I don't think it's readable.

p.s. Both Java and JavaScript allows having labels on break and continue so having this in C# would come in handy at times.

**Comment by DavidArno** [comment]
@jnm2,
```cs
(bool found, ValueType value) TryFindValueIn2DStructure(...
```
I've never really been convinced by the "naming is hard" argument. If you understand what a piece of code does, you can describe it. If you can describe it, you can summarise it. That summary gives you the name. If the summary is long (and likely contains lots of and's or then's), then it's probably doing too much and could do with breaking into smaller parts.

**Comment by eyalsk** [comment]
@DavidArno "naming is hard" because thinking about it objectively is indeed hard, different people may have different opinions about how something should be named and it's even harder for a name to make sense outside to the few lines of code you or someone else wrote and last but not least code tend to change over time and name things so everything will make sense all of the time, especially the future isn't that simple.

**Comment by HaloFour** [comment]
I agree with @DavidArno, if you have a need to break from a loop then you have a reason for wanting to do so.  It shouldn't be that difficult to summarize that into a quick one or two word label, e.g. `FOUND_GROMIT` (if you're into the whole upper-case label convention).  What's more this is self-describing so other developers (or yourself in six months) should be able to infer what you're doing and why you did it from a glance.

**Comment by Joe4evr** [comment]
[Also related.](https://github.com/dotnet/roslyn/issues/5883#issuecomment-153798232)

**Comment by tpetrina** [comment]
Maybe immediately invoked functions with return could help with the naming problem...

**Comment by eyalsk** [comment]
Just to clarify, my comment earlier was in regard to "There are only two hard things in Computer Science: cache invalidation and naming things." and really the way I think about the word hard in this context is you need to **think** hard about naming stuff but yeah probably, if you can explain a piece of code as @DavidArno said then one can come up with a reasonable name, it might not satisfy everyone but at least it would be reasonable.

**Comment by gulshan** [comment]
Can I put the label and loop from @HaloFour 's example in the same line?
```csharp
String foundValue = null;
OUTERLOOP: for (int x = 0; x < xMax; x++)
{
    for (int y = 0; y < yMax; y++)
    {
        foundValue = GetValue(x, y);
        if (foundValue == valueToCompare)
            break OUTERLOOP;
    }
}
```

**Comment by eyalsk** [comment]
@gulshan In JavaScript it's allowed, I'm not sure about Java though, haven't used this feature there.

**Comment by HaloFour** [comment]
@gulshan Yes

**Comment by orthoxerox** [comment]
The VB LDT has positively reviewed a similar feature where the loop is referenced not by its level or its label, but by its loop variable. This won't work for C#, as only `foreach` is guaranteed to have a loop variable.

**Comment by gafter** [comment]
`foreach` is no longer guaranteed to have a loop variable.
1. `foreach (var _ in e) ...`
2. `foreach (var (x, y) in e) ...`

**Comment by juliusfriedman** [comment]
Yes... make a LoopState or LoopContext and refer to that upon exit.

It doesn't have to be complex but the compiler should not be doing things like this because the question becomes when where and how not Why and it's why that always bites.

Ntl.

Break from within another switch or while now has different semantics especially if performed at the end of a loop...

I vote no.

**Comment by bmuessig** [comment]
It's a dearly lacking feature. I vote yes.

**Comment by jez9999** [comment]
I fully support implementing labeled loops.  They would be relatively easy to implement (basically syntactic sugar) and look a bit less ugly than using `goto`:

```
outerloop:
while (...) {
    while (...) {
        if (...) { continue outerloop; }
        break outerloop;
    }
}
```

Rather than:

```
while (...) {
    while (...) {
        if (...) { goto continue_outerloop; }
        goto break_outerloop;
    }

    continue_outerloop: {}
}
break_outerloop: {}
```

**Comment by troien** [comment]
I like this idea as I really want this to be simpler then it is now, but personally I would prefer a different syntax.

`break break;`
`break break break;`
`break continue; // this would break out of current loop/switch and continue in 'parent' loop`

Where you can basically add as many 'break' as you need, but there can only be one continue and continue has to be the last. (as it wouldn't make sense to continue and then immediatly break)

```
while(...)
{
    while(...)
   {
      if (...) break break;
   }
}
```

```
while(...)
{
    if(...)
        break break; // should give compiler error CS0139
}
```
The advantage of this would be that the break statements are still linked to a loop at compile time, meaning the compiler can treat them as if they were placed in the enclosing loop (or switch), show compiler errors and when putting your cursor on a break statement, it should still be able to highlight the correct loop just like it does now.

**Comment by marcussacana** [comment]
I discussed about this feature as well in the [vs dev community](https://developercommunity.visualstudio.com/idea/437245/break-multiple-loops.html) 2 years ago.
Nothing implemented yet but should be good see implemented one day,
To me `break 2;` or `break label_name;` looks good.

**Comment by merarischroeder** [comment]
I noticed today that Rust has this kind of thing: `break 'outer;` - it's nice! see https://doc.rust-lang.org/rust-by-example/flow_control/loop/nested.html

I came looking to see if C# has this yet. I'm hoping it comes along one day.

So Java and Rust have this kind of language feature. Are there any other languages with this apart from Java and Rust?

(Rust uses a dashed syntax for loop labels, because they also support Loop Expressions - https://doc.rust-lang.org/reference/expressions/loop-expr.html#continue-expressions)

**Comment by TahirAhmadov** [comment]
I, too, would support labeled `break` statements similar to Java and Rust. Existing label syntax should be used to "name" `while`, `for`, `foreach`, `switch`, `do` constructs; these labels can `goto`ed, `break`ed, and `continue`d.

**Comment by leoshusar** [comment]
+1 for labels for me.

For example Dart has also labels like these:
```dart
outer: for (final a in [1, 2, 3]) {
  inner: for (final b in ['A', 'B', 'C']) {
    if (b == 'C')
      continue inner;
    else if (a == 3)
      break outer;
    print ('$a $b');
  }
}
```

and Kotlin even allows to return from lambda functions like `forEach`:
```kt
listOf(1, 2, 3, 4, 5, 6).forEach loop@{
    if (it % 2 == 0) return@loop
    print(it)
}
// prints 135
```
but here you can also omit your label and use implicit label, like `return@forEach`

**Comment by TheBlackPlague** [comment]
From what I can infer, most people agree that some sort of in-context multi-level breaking syntax is needed. In which case, why can't we change the label on this to `Proposal`? @jcouv / @mattwar

**Comment by DavidArno** [comment]
@TheBlackPlague,

Since the OP for this discussion has 39 upvotes and 30 downvotes, I'd suggest that "*most people agree that some sort of in-context multi-level breaking syntax is needed*" is a strange conclusion to reach here. 😄

**Comment by jez9999** [comment]
I don't know why anyone would downvote this.  I mean, what skin is it off anyone's nose if it gets implemented and they don't like it?  Just don't use it.  Sadly, owing to the indifference and irrational hostility to the proposal in the C# community that seems to exist I've given up hope of it ever being implemented, which sucks, because I'm forever going to be using labels pointing to empty blocks along with `goto`s to achieve the same result.

**Comment by CyrusNajmabadi** [comment]
> I don't know why anyone would downvote this as a general idea.

People are likely not down voting the general idea, but rather the specific solution outlined in the OP.

**Comment by CyrusNajmabadi** [comment]
> Sadly, owing to the indifference and irrational hostility 

Painting those who disagree with you as irrational is a violation of the .net code if conduct. As we say there:

> Examples of behavior that contributes to a positive environment for our community include:

> Demonstrating empathy and kindness toward other people
> Being respectful of differing opinions, viewpoints, and experiences
> Giving and gracefully accepting constructive feedback

**Comment by DavidArno** [comment]
> I don't know why anyone would downvote this as a general idea

I can only speak for myself, but I actively downvote all proposals around labelled breaks as I strongly dislike the general idea.

The reason I dislike it is because it removes the pain involved in handling deeply nested conditionals. There's reams of information on the web around why deep nesting is a bad idea with plenty of advice on refactoring such code. With that in mind, C# current approach of requiring the use of `goto` to break out of deep nesting is an example of "pit of success" functionality. You have to work hard at writing code that climbs out of the success pit. Adding labelled breaks makes it easier to write deeply nested code, which goes against that pit of success principle and so I voice my opposition to any such change.

**Comment by marcussacana** [comment]
If I want to work hard for simple tasks I would be using assembly

**Comment by NN---** [comment]
A wider solution is to adapt [Nemerle block](https://github.com/rsdn/nemerle/wiki/Block) which allows doing this and more.
It makes possible to define a loop as expression adding an ability to write code without lambda capturing or explicit local function.

```cs
string a = x: {
 foreach (var q in w) if (q == 1) x("found 1"));
 x("not found"));
};
```

**Comment by pm100** [comment]
came here looking for labeled loops that can be referred to in break or continue. I was sure the feature existed but seems not (must have been thinking of java, rust and go)

**Comment by chucker** [comment]
Swift has labeled loops and a break label syntax for this (as do apparently Rust and others). It’s basically just syntactic sugar for the goto approach, but I think it’s worth considering.

Swift example: https://twitter.com/zet_manu/status/1498324614346276866

**Comment by NigelBess** [comment]
> Small focused functions are your friend. But if you want long functions, as you say, there is already two ways of achieving this goal. So I'd see yet another way of achieving it as really bringing nothing to the language.

Sometimes you inherit a monstrous legacy code base and refactoring into small functions is not an option. For these cases, `break N` would be really useful.

"C# was invented by pragmatic programmers who have jobs to do, for pragmatic programmers who have jobs to do." - Eric Lippert

**Comment by CyrusNajmabadi** [comment]
@NigelBess Can you just use 'goto' in that case?

**Comment by leoshusar** [comment]
`goto` is more error-prone than simple `break outer`.

**Comment by CyrusNajmabadi** [comment]
@leoshusar how so?

**Comment by leoshusar** [comment]
```dart
outer: for (final a in [1, 2, 3]) {
  inner: for (final b in ['A', 'B', 'C']) {
    if (b == 'C')
      continue inner;
    else if (a == 3)
      break outer;
    print ('$a $b');
  }
}
```
vs
```dart
for (final a in [1, 2, 3]) {
  for (final b in ['A', 'B', 'C']) {
    if (b == 'C')
      goto inner;
    else if (a == 3)
      goto outer;
    print ('$a $b');
    inner:
  }
}
outer:
```

Which one is more clear?

Also with the `goto` statement someone can put something after the `inner:` label or before the `outer:` label, breaking your `continue` and `break` intentions.

**Comment by CyrusNajmabadi** [comment]
> Which one is more clear?

Definitely the latter for me.

> Also with the goto statement someone can put something after the inner: label or before the outer: label, breaking your continue and break intentions.

I'm not sure why that matters.  Having more flexibility seems totally fine here.  And, if you don't want that, just write an analyzer that disallows that if htat's your personally preferred style.

In the context of the discussion i was responding to, this was about "inherit[ing] a monstrous legacy code base".  I don't see what's wrong with using 'goto' there. And i def think it's reasonable to create analyzers to find and address issues you find in that legacy codebase.

**Comment by CyrusNajmabadi** [comment]
> This would make it slightly safer than goto.

What do you mean by 'safer than goto'?  What is unsafe about goto?

**Comment by CyrusNajmabadi** [comment]
I am definitely not a fan of introducing a more restrictive lang feature just because someone thinks the existing capability is too permissive.  If the feeling is that something is too permissive, then that much better solved with an analyzer vs adding an entire new feature to just provide a subset of existing functionality.

**Comment by leoshusar** [comment]
> Definitely the latter for me.

Interesting, I look at the code and when I see `continue inner`, I immediately know it means the inner loop should continue.
When I look at the latter, I don't see it when I see the `goto`, I have to look where that jump label is and why is it at the end of the loop.

> I am definitely not a fan of introducing a more restrictive lang feature just because someone thinks the existing capability is too permissive.

But nobody takes away the `goto`, you can freely continue using it. There are many people who would appreciate this feature, the rest doesn't have to use it. Also considering how many other languages have this.
The same could be said about kinda every feature - why have auto properties? They are too permissive, just create backing field 
and getter / setter methods for it, etc.

"Why take this feature away just because someone thinks it's unnecessary for him?"

**Comment by CyrusNajmabadi** [comment]
?

My argument is the opposite.  We already have a permissive feature, so we shouldn't create restrictive features to duplicate them.  You can just use an analyzer if you don't like the idea of having non-loop code follow the goto. 

>But nobody takes away the goto, you can freely continue using it. 

Right... that's my point.  Since goto works well for this, and is already tehre, and already supports a superset of this functionality with great syntax, i would stick with it.  If that superset is not desirable, i would recommend using an analyzer versus just creating a duplicated feature here.

**Comment by CyrusNajmabadi** [comment]
>  There are many people who would appreciate this feature,

**Comment by macias** [comment]
> What do you mean by 'safer than goto'? What is unsafe about goto?

You write your code basically once, but you maintain it for a long time -- it means you read it more then you write it. So it is crucial the code speaks to you in clear fashion -- goto does not have any special meaning, break/continue has. If someone misplace/misuse the label it is harder to find the bug than with break/continue because the latter have extra information embedded in them -- they are telling the reader the meaning.

Or, shorter version, goto was from the beginning, so why we have break/continue in the first place, if we could live just with goto?

**Comment by CyrusNajmabadi** [comment]
> goto does not have any special meaning

I mean, it means: go to this label :)

> If someone misplace/misuse the label

What sort of misuse are you thinking of?  If you're worried about placement, that ties into what i was saying above, just add an analyzer to restrict to only teh stylistic uses of goto you're comfortable with.

> goto was from the beginning, so why we have break/continue in the first place

Because they don't need labels.  If you need labels, at that point, just use 'goto' :)

**Comment by leoshusar** [comment]
@CyrusNajmabadi Sorry, non-English, confused restrictive with permissive :)

But how would you write an analyzer that knows if you want to `goto` as a `goto` or use `goto` as a `break` or `continue`? What if you want to use both in your code? I don't see any easy way to do this.

Also I've written so far 2 super simple analyzers and I don't expect majority of people invest time to learn Roslyn "just for this".

**Comment by TahirAhmadov** [comment]
Would it be better to use keywords instead of labels?
```cs
while(...)
{
  while(...)
  {
    break outer;
    while(...)
    {
      break outer outer;
    }
  }
}
```

**Comment by leoshusar** [comment]
@TahirAhmadov what if someone adds another loop? He would have to remember to add another `outer` to all inner breaks.
In my opinion this is the same as already proposed `break 3` which I personally don't like. I see this even more error-prone than `goto`.

**Comment by CyrusNajmabadi** [comment]
>Also I've written so far 2 super simple analyzers and I don't expect majority of people invest time to learn Roslyn "just for this".

The majority of people don't need to do this.  Someone just needs to write it once and make into a nuget package.  If users want this sort of linting, they can choose to use that package.  If htey don't care, then there's no issue :)

If users care, but not a single person cares enough to write an analzyer, then it sounds like there isn't much caring after all, in which case i'd def believe there's not a great reason to make this into a lang feature.

**Comment by leoshusar** [comment]
But still, analyzer would solve an issue for people who don't like `goto` at all.
My point is, what if someone wants to use both `goto` and safer loop breaking? What would that analyzer look like?

Also this still wouldn't solve the readability issue, which.. for you is fine, but I would definitely see the loop labels more readable, especially for more complex code than 5 line example.

**Comment by CyrusNajmabadi** [comment]
> and safer loop breaking? 

I'm still not sure why it's safer tbh.

> What would that analyzer look like?

Maybe say that if a goto is in a loop, it needs to be at the end of the loop.  If you want inner gotos and loop gotos, then you have to decide what you want.

**Comment by leoshusar** [comment]
People, including me, have already written their opinions on why `goto` is more error-prone.

> then you have to decide what you want

You see, analyzer doesn't solve everything this feature would solve. You would have to have either one or the other, not both.

**Comment by CyrusNajmabadi** [comment]
> People, including me, have already written their opinions on why goto is more error-prone.

I don't really get your argument.  All you said was: "Interesting, I look at the code and when I see continue inner, I immediately know it means the inner loop should continue."

You can't know that.  You have to go figure that out.  For example, the code coudl have said:

```c#
inner: foreach (...) {
    outer: foreach (...) {
```

now, the `continue inner` doesn't mean the inner loop continues, it actually means the outer one continues.  You have to go find the starts of the regions to figure this out.  This same is true with goto, it's just the ends.  And the ends make *much* more sense to me as that's where execution actually goes to.  For example with continue, while the label precedes the loop, that's now wher eyou're jumping to.  You're actually jumping to the next *iteration*.  Which to me is clearly indicated by a common point (the loop end).  Indeed, that's what it means to hit the endpoint of a loop, that you continue on the next iteration.  

Labeling this explicitly and then stating you want to go there makes the most sense to me and actually fits what the language is doing here.  It was also so much clearer in the code you had in your example. Indeed with your example code it was much more confusing as now i had to say "oh... i have to be careful if someone goto's this location, as opposed to 'continue'ing to it".

> You see, analyzer doesn't solve everything this feature would solve.

If the cases it doesn't solve are so niche, then that's not something that makes me want to add this feature.

If the only reason to have htis feature is so that a subset of users can mix-match gotos and continues/breaks in a loop and have differing semantics for them, then that's so minor that i don't think it's warranted as a language feature.

**Comment by leoshusar** [comment]
I also said this:

> Also with the goto statement someone can put something after the `inner:` label or before the `outer:` label, breaking your `continue` and `break` intentions.


With the `break responseLoop` I'm saying "I want to break the `responseLoop`". And I bet people who are familiar with `break` and `continue` would 100% get what that means.

Unlike the `goto`, you would need to have both `breakResponseLoop` and `continueResponseLoop` and ensure they are really in the right place.

Now you can say "so what?" and I agree that this is not a critical feature that the majority of C# programmers will use, but why not have it for the rest? Other languages have it and I'm sure it has it's users. For example Golang is "fairly new" and has both `goto` and labeled loops. I think they wouldn't add it if they knew nobody would use it.

**Comment by SirDragonClaw** [comment]
> I am definitely not a fan of introducing a more restrictive lang feature just because someone thinks the existing capability is too permissive.

This ignores the fact that GOTO is so permissive that it's use has been banned from the code bases of the last 4 places I have worked at. GOTO and its use is very frowned upon, for many codebases this would be a cleaner replacement, but I understand the downsides too. 

That being said I find it amusing you are all still discussing this issue 5 years after I first submitted the proposal 😂

**Comment by marcussacana** [comment]
> That being said I find it amusing you are all still discussing this issue 5 years after I [first submitted](https://github.com/dotnet/roslyn/issues/21864) the proposal joy

Well, I'm here too, and this topic is even 'more or less' the continuation from the old visual studio forum, so it's more than 5 years.  
Anyway, to me, feature is feature, the newer C# versions implements some features that I never even really used one time in my life, and maybe I will never use it, but even with that I'm not saying that feature should't be implemented.

We are programmers, not kids, if the goto is good or bad to the code should be the programmer should decide this.  
If you or your friends don't need a feature you can just ignore it.

C# is a high level language, so I should aim to simplify the programmer's life where I can, as such I find this feature interesting, and frankly, having to fill the recursive loops relating to the previous loop with booleans and ifs, in my opinion is what makes confusing code, many cases where the code continuation point could be executed where it was necessary to decide where it needs to continue could be solved with this, instead of creating new variables and ifs to check every each loop.  
And I find it relatively less confusing than a sequence of yield break/continue loops.

**Comment by theunrepentantgeek** [comment]
@NigelBess wrote:
> Sometimes you inherit a monstrous legacy code base and refactoring into small functions is not an option. 

and @leoshusar responded:
> But nobody takes away the **goto,** you can freely continue using it. There are many people who would appreciate this feature, the rest doesn't have to use it. 

There's a fallacy here - even a moderate sized project (say, 200kloc) will have multiple developers and maintainers and a lifespan measured in years.

If _even one_ of those developers uses a particular C# feature, then _all_ of those developers needs to be aware of it, else they'll run into code they cannot reason about. The developer who introduced use of the feature may have retired and be living off the grid in an island paradise by the time a new developer runs into the code.

Saying "if people don't like it, they don't have to use it" ignores the reality that most business development involves teams working on a common codebase over several years.

**Comment by CyrusNajmabadi** [comment]
> GOTO and its use is very frowned upon

Why is it frowned upon?

**Comment by SirDragonClaw** [comment]
> > GOTO and its use is very frowned upon
> 
> Why is it frowned upon?

https://www.l3harrisgeospatial.com/Learn/Blogs/Blog-Details/ArtMID/10198/ArticleID/15289/Why-should-GOTO-be-avoided

This was the first result in Google, there are 1000's more. Personally I think it has its uses and banning it from codebases is going to far, but many disagree.

**Comment by jez9999** [comment]
Boy do I with MS would introduce labeled loops.  It's literally the one totally valid use-case for `goto`.  And yet they always tell people not to use `goto`.

Apparently new C# feautres need a 'champion'.  Well, I'm not sure why nobody seems to be willing to champion this considering I've seen comments from hundreds of developers who want it.

I'll also just mark here that a related discussion is going on at https://github.com/dotnet/csharplang/discussions/5525

**Comment by HaloFour** [comment]
> And yet they always tell people not to use `goto`.

Because it's not the C# language team telling people to not use `goto`, they still feel that it's completely appropriate to use that in these cases.  The `goto` keyword in C# is pretty constrained and mitigates most of the issues that led to that book being written over 50 years ago.  Even Djiksta didn't think the advice should be followed dogmatically.
